File: Programação III/TrabalhoJava/src/Eleicao.java
import java.util.Date;
import java.util.LinkedList;
import java.text.DecimalFormat;
import java.text.SimpleDateFormat;

public class Eleicao {
    private LinkedList<Candidato> candidatosEleitos;
    private LinkedList<Candidato> candidatosMaisVotados;
	private LinkedList<Partido> partidos;
	private LinkedList<Partido> partidosPrimUlt;
	private Integer[] idades = {0, 0, 0, 0, 0};
	private Integer[] sexos = {0, 0}; 
	private Integer numVagas = 0;
	private Integer votosNominais = 0;
	private Integer votosLegenda = 0;
 

	public Integer getNumVagas() {
		return this.numVagas;
	}

	public LinkedList<Candidato> getCandidatosEleitos() {
		return this.candidatosEleitos;
	}

	public void setCandidatosEleitos(LinkedList<Candidato> candidatos, String data) {
		this.candidatosEleitos = new LinkedList<Candidato>();

		SimpleDateFormat formato = new SimpleDateFormat("dd/MM/yyyy"); 
		Date dataEleicao = null;
		try {
			dataEleicao = formato.parse(data);
			
		}catch(Exception e){
			System.out.println(e);
		}
        
        for(Candidato candidato: candidatos){
            if("Eleito".equals(candidato.getSituacao())){                              
                candidatosEleitos.add(candidato);
				this.numVagas++;
				
				Integer idade = candidato.getIdade(dataEleicao);

				String sexo = candidato.getSexo();
				if("M".equals(sexo)){
					sexos[0]++;
				}
				else{
					sexos[1]++;
				}

				if(idade < 30){
					idades[0]++;
				}
				else if(idade >= 30 && idade < 40){
					idades[1]++;
				}
				else if(idade >= 40 && idade < 50){
					idades[2]++;
				}
				else if(idade >= 50 && idade< 60){
					idades[3]++;
				}
				else {
					idades[4]++;
				}
			}
        }
	}

	public LinkedList<Candidato> getCandidatosMaisVotados() {
		return this.candidatosMaisVotados;
	}

	public void setCandidatosMaisVotados(LinkedList<Candidato> candidatos) {
		this.candidatosMaisVotados = new LinkedList<Candidato>();

		Integer i = 0;
		
		for(Candidato candidato: candidatos){
            candidatosMaisVotados.add(candidato);
			i++;

			if("Válido".equals(candidato.getDestinoVoto())){
				this.votosNominais += Integer.parseInt(candidato.getVotosNominais());
			}
        }
	}

	public LinkedList<Partido> getPartidos() {
		return this.partidos;
	}

	public void setPartidos(LinkedList<Partido> partidos) {
		this.partidos = new LinkedList<Partido>();
		this.partidos.addAll(partidos);
	}

	public LinkedList<Partido> getPartidosPrimUlt() {
		return this.partidosPrimUlt;
	}

	public void setPartidosPrimUlt(LinkedList<Partido> partidos) {
		this.partidosPrimUlt = new LinkedList<Partido>();
		this.partidosPrimUlt.addAll(partidos);
	}

    public void imprimeCandidatosEleitos(){
		System.out.println("Vereadores eleitos:");
		Integer i = 1;
        for(Candidato candidato: this.candidatosEleitos){
			System.out.print(i + " - ");
            System.out.println(candidato);
			i++;
        }
	}

    public void imprimeCandidatosMaisVotados(){
		System.out.println("\nCandidatos mais votados (em ordem decrescente de votação e respeitando número de vagas):");
        Integer i = 1;
		if(this.numVagas == 0){
			return;
		}
		for(Candidato candidato: this.candidatosMaisVotados){
			System.out.print(i + " - ");
            System.out.println(candidato);
			i++;

			if(i > this.numVagas){
				break;
			}
        }
	}

	public void imprimeEleitosMajoritario(){
		Integer i = 1;
		Integer contador=0;
		System.out.println("\nTeriam sido eleitos se a votação fosse majoritária, e não foram eleitos:\n(com sua posição no ranking de mais votados)");
		for(Candidato candidato: this.candidatosMaisVotados){
			Integer existe=0;
			
			for(Candidato candidatoEleito: this.candidatosEleitos){
				if(candidato.comparaCandidatoNome(candidatoEleito)==1){
					existe = 1;	
					contador++;
				}
			}
			if((existe==0 ) &&( contador < this.numVagas)){
			 System.out.print(i + " - ");
             System.out.println(candidato);
			 contador++;
			}
			existe=0;
			i++;
        }
	}


    public void imprimeEleitosQueSeBeneficiaram(){
		Integer i=1;
		System.out.println("\nEleitos, que se beneficiaram do sistema proporcional:\n(com sua posição no ranking de mais votados)");
		for(Candidato candidato: this.candidatosMaisVotados){
			if(i > this.numVagas){
				if("Eleito".equals(candidato.getSituacao())){            
					System.out.print(i + " - ");              
					System.out.println(candidato);
				}
			}
			i++;
        }
	}


	public void imprimeVotacaoPartidos(){
		Integer i = 1;
		System.out.println("\nVotação dos partidos e número de candidatos eleitos:");
		for(Partido partido: this.partidos){
			System.out.print(i + " - ");   
			System.out.println(partido);
			i++;


			this.votosLegenda += Integer.parseInt(partido.getVotosLegenda());
        }
	}

	public void imprimePartidosPrimUlt(){
		Integer i = 1;
		System.out.println("\nPrimeiro e último colocados de cada partido:");
		for(Partido partido: this.partidosPrimUlt){
			if(partido.getVotosTotais() != 0){
				System.out.print(i + " - ");   
				partido.imprimePartidoPrimUlt();
				i++;
			}
        }
	}

	public void filtroVotosNominaisPartido(LinkedList<Partido> partidos){
		LinkedList<Partido> nova = new LinkedList<Partido>();
    
		for(Partido partido: partidos){
			if(partido.getVotosNominais() > 0){
				nova.add(partido);
			}
		}

		 this.setPartidosPrimUlt(nova); 	
	}



	private String calculaPorcentagem(Integer quant){
		Float result = (quant.floatValue()/this.numVagas.floatValue())* 100;
		return new DecimalFormat("#,##0.00").format(result);
	}

	public void imprimePorFaixaEtaria(){
		String porcentagem = "0.00";
		System.out.println("\nEleitos, por faixa etária (na data da eleição):");

		porcentagem = calculaPorcentagem(this.idades[0]);
		System.out.println("      Idade < 30: " + idades[0] +" ("+ porcentagem +"%)");

		porcentagem = calculaPorcentagem(this.idades[1]);
		System.out.println("30 <= Idade < 40: " + idades[1] +" ("+ porcentagem +"%)");

		porcentagem = calculaPorcentagem(this.idades[2]);
		System.out.println("40 <= Idade < 50: " + idades[2] +" ("+ porcentagem +"%)");

		porcentagem = calculaPorcentagem(this.idades[3]);
		System.out.println("50 <= Idade < 60: " + idades[3] +" ("+ porcentagem +"%)");

		porcentagem = calculaPorcentagem(this.idades[4]);
		System.out.println("60 <= Idade     : " + idades[4] +" ("+ porcentagem +"%)");
	}

	public void imprimePorSexo(){
		String porcentagem = "0.00";
		System.out.println("\nEleitos, por sexo:");

		porcentagem = calculaPorcentagem(this.sexos[1]);
		System.out.println("Feminino:  " + sexos[1] +" ("+ porcentagem +"%)");

		porcentagem = calculaPorcentagem(this.sexos[0]);
		System.out.println("Masculino: " + sexos[0] +" ("+ porcentagem +"%)");
	}

	private String calculaPorcentagemTotal(Integer total, Integer valor){
		Float result = (valor.floatValue()/total.floatValue())* 100;
		return new DecimalFormat("#,##0.00").format(result);
	}

	public void imprimeVotosTotais(){
		Integer votosValidos = this.votosNominais + this.votosLegenda;
		String porcentagem = "0.00";

		System.out.println("\nTotal de votos válidos:    " + votosValidos);

		porcentagem = calculaPorcentagemTotal(votosValidos, this.votosNominais);
		System.out.println("Total de votos nominais:   " + this.votosNominais + " (" + porcentagem + "%)");
		
		porcentagem = calculaPorcentagemTotal(votosValidos, this.votosLegenda);
		System.out.println("Total de votos de Legenda: " + this.votosLegenda + " (" + porcentagem + "%)\n");
	}
}

File: Programação III/TrabalhoCpp/src/Eleicao.cpp
#include "Eleicao.h"
#include "NumberUtils.h"
using namespace std;

Eleicao::Eleicao() {
    this->votos_nominais = 0;
    this->votos_legenda = 0;

    for(int i = 0; i < 5; i++){
        idades[i] = 0; 
        if(i < 2)
        sexos[i] = 0;
    }

}

static bool compare_to_candidato_mais_votados(Candidato* first , Candidato* second){

    if (stoi(first->getVotosNominais()) > stoi(second->getVotosNominais())) { 
        return true; 
    } 
    else if(stoi(first->getVotosNominais()) < stoi(second->getVotosNominais())){ 
        return false;
    }
    else{
        return first->getIdade() > second->getIdade();
    }


}

void Eleicao::cria_lista_candidatos(list<Candidato*>& lista_candidatos){
    lista_candidatos.sort(compare_to_candidato_mais_votados);  
    this->lista_candidatos = lista_candidatos;
}


void Eleicao::cria_lista_candidatos_eleitos(const list<Candidato*>& lista_candidatos){
    for (auto& candidato : lista_candidatos){
        if(candidato->candidato_eleito()){
            this->lista_candidatos_eleitos.push_back(candidato);
            this->num_vagas++;

            int idade = calcula_idade(candidato->getDataNasc() , "15/11/2020");
            string sexo = candidato->getSexo();
            if("M" == sexo){
                sexos[0]++;
            }
            else{
                sexos[1]++;
            }

            if(idade < 30){
                idades[0]++;
            }
            else if(idade >= 30 && idade < 40){
                idades[1]++;
            }
            else if(idade >= 40 && idade < 50){
                idades[2]++;
            }
            else if(idade >= 50 && idade< 60){
                idades[3]++;
            }
            else {
                idades[4]++;
            }
        }
    }
}
//Essa função não tem retorno e recebe um ponteiro para uma lista de candidatos
//tem o objetivo de criar a lista de candidatos mais votados 
void Eleicao::cria_lista_candidatos_mais_votados(const list<Candidato*>& lista_candidatos){
    for (auto& candidato : lista_candidatos){
        this->lista_candidatos_mais_votados.push_back(candidato);

        string destino_voto = candidato->getDestinoVoto();
        if(destino_voto.compare("Válido") == 0){   //
            this->votos_nominais += stoi(candidato->getVotosNominais());
        }
    }
}

 void Eleicao::imprime_lista_candidatos_eleitos_beneficiaram(){
    cout << "Eleitos, que se beneficiaram do sistema proporcional:" << endl;
    cout << "(com sua posição no ranking de mais votados)" << endl;
     int i = 1;
     for (auto& candidato : this->lista_candidatos_mais_votados){
         if(i > this->num_vagas){
            string situacao = candidato->getSituacao();
            if(situacao.compare("Eleito") == 0){
                cout << i << " - ";
                candidato->println();
            }
         }
         i++;
     }
     cout << endl;
 }


void Eleicao::imprime_lista_candidatos_eleitos(){
    cout << "Número de vagas: " << this->num_vagas << endl << endl;
    cout << "Vereadores eleitos:" << endl;

    int i = 1;
    for (auto& candidato : this->lista_candidatos_eleitos){
        cout << i << " - ";
        candidato->println();
        i++;
    }
    cout << endl;
}

void Eleicao::imprime_lista_candidatos_mais_votados(){
    cout << "Candidatos mais votados (em ordem decrescente de votação e respeitando número de vagas):" << endl;

    int i = 1;
    for (auto& candidato : this->lista_candidatos_mais_votados){
        if(i > this->num_vagas){
            break;
        }
        cout << i << " - ";
        candidato->println();
        i++;
    }
    cout << endl;
}


void Eleicao::imprime_lista_candidatos_mais_votados_nao_eleitos(){
    cout << "Teriam sido eleitos se a votação fosse majoritária, e não foram eleitos:" << endl;
    cout << "(com sua posição no ranking de mais votados)" << endl;
    int i = 1, j = 1;
    for (auto& candidato : this->lista_candidatos_mais_votados){
        if(i > this->num_vagas){
            break;
        }
        string situacao = candidato->getSituacao();
        if(situacao.compare("Eleito")  != 0){
            cout << j << " - ";
            candidato->println();
        }
        j++;
        i++;
    }
    cout << endl;
} 


void Eleicao::cria_lista_partidos(const list<Partido*>& lista_partidos) {
    this->lista_partidos = lista_partidos;
}


static bool ordena_partidos(Partido* first , Partido* second){
    if (first->get_votos_totais() > second->get_votos_totais()) { 
        return true; 
    } 
    else { 
        return false;
    }
}


static bool ordena_partidos_prim_ult(Partido* first , Partido* second){

    if(first->get_numero_candidatos_partido() < 1){
        // cout << "aqui 1 ----------------------------------" << endl;
        return true;
    } 
    else if(second->get_numero_candidatos_partido() < 1){
        // cout << "aqui 2 ----------------------------------" << endl;
        return false;
    }

    Candidato* candidato_prim = first->get_primeiro_candidato(); 
    Candidato* candidato_sec = second->get_primeiro_candidato();
    
    int votos_first = stoi(candidato_prim->getVotosNominais());
    int votos_second = stoi(candidato_sec->getVotosNominais());

    if (votos_first > votos_second){ 
        return true;
    } 
    if(votos_first < votos_second){ 
        return false;
    }
    
    return (stoi(candidato_prim->getNumeroPartido()) < stoi(candidato_sec->getNumeroPartido())); 
 }

void Eleicao::imprime_lista_partidos(){
     cout << "Votação dos partidos e número de candidatos eleitos:" << endl;
     this->lista_partidos.sort(ordena_partidos);
     int i = 1;
     for (auto& partido: this->lista_partidos){
        cout << i << " - ";
        partido->println();
        this->votos_legenda += stoi(partido->get_votos_legenda());
        i++;
    } 
    cout << endl;
}


void Eleicao::imprime_lista_partidos_prim_ult(){
    this->lista_partidos.sort(ordena_partidos_prim_ult);
   
    cout << "Primeiro e último colocados de cada partido:" << endl;
    int i = 1;
    for (auto& partido: this->lista_partidos){
        if(partido->get_votos_totais() != 0){
            cout << i << " - ";
            partido->imprime_partido_prim_ult();
            i++;
        }
    } 
    cout << endl;
}

double Eleicao::calcula_porcentagem(int quant){
    double result = (quant*1.0/this->num_vagas)* 100;
    return result;
}

int Eleicao::calcula_idade(const string& data_cand , const string& data_eleicao){
    int dia_cand = stoi(data_cand.substr(0,2));
    int mes_cand = stoi(data_cand.substr(3,2));
    int ano_cand = stoi(data_cand.substr(6,4));

    int dia_eleicao = stoi(data_eleicao.substr(0,2));
    int mes_eleicao = stoi(data_eleicao.substr(3,2));
    int ano_eleicao = stoi(data_eleicao.substr(6,4));

    int idade;
    if (mes_eleicao > mes_cand || (mes_eleicao == mes_cand  && dia_eleicao >= dia_cand)) {
        idade = ano_eleicao - ano_cand;
    }else {
     idade = ano_eleicao - ano_cand - 1;
    }
    return idade;

}


void Eleicao::imprime_por_faixa_etaria(){
    double porcentagem = 0.00;

    cout << "Eleitos, por faixa etária (na data da eleição):" << endl;

    porcentagem = calcula_porcentagem(this->idades[0]);
    cout << "      Idade < 30: " << idades[0] << " (" << formatDoubleCurrency(porcentagem, LOCALE_PT_BR) << "%)" << endl;

    porcentagem = calcula_porcentagem(this->idades[1]);
    cout << "30 <= Idade < 40: " << idades[1] << " (" << formatDoubleCurrency(porcentagem, LOCALE_PT_BR) << "%)" << endl;

    porcentagem = calcula_porcentagem(this->idades[2]);
    cout << "40 <= Idade < 50: " << idades[2] << " (" << formatDoubleCurrency(porcentagem, LOCALE_PT_BR) << "%)" << endl;

    porcentagem = calcula_porcentagem(this->idades[3]);
    cout << "50 <= Idade < 60: " << idades[3] << " (" << formatDoubleCurrency(porcentagem, LOCALE_PT_BR) << "%)" << endl;

    porcentagem = calcula_porcentagem(this->idades[4]);
    cout << "60 <= Idade     : " << idades[4] << " (" << formatDoubleCurrency(porcentagem, LOCALE_PT_BR) << "%)\n" << endl;

}

void Eleicao::imprime_por_sexo(){
    double porcentagem = 0.00;

    cout << "Eleitos, por sexo:" << endl;

    porcentagem = calcula_porcentagem(this->sexos[1]);
    cout << "Feminino:  " << sexos[1] << " (" << formatDoubleCurrency(porcentagem, LOCALE_PT_BR) << "%)" << endl;

    porcentagem = calcula_porcentagem(this->sexos[0]);
    cout << "Masculino: " << sexos[0] << " (" << formatDoubleCurrency(porcentagem, LOCALE_PT_BR) << "%)\n" << endl;

}


void Eleicao::imprime_votos_totais(){
    double porcentagem = 0.00;

    int votos_totais = this->votos_legenda + this->votos_nominais;
    
    cout << "Total de votos válidos:    " << votos_totais << endl;

    porcentagem = (this->votos_nominais*1.0/votos_totais)* 100;
    cout << "Total de votos nominais:   " << this->votos_nominais << " (" << formatDoubleCurrency(porcentagem, LOCALE_PT_BR) << "%)" << endl;

    porcentagem = (this->votos_legenda*1.0/votos_totais)* 100;
    cout << "Total de votos de Legenda: " << this->votos_legenda << " (" << formatDoubleCurrency(porcentagem, LOCALE_PT_BR) << "%)\n" << endl;

}




Eleicao::~Eleicao(){
    for (auto& candidato : this->lista_candidatos){
        delete candidato;
    }

    for (auto& partido : this->lista_partidos){
        delete partido;
    }
}